# Améliorez le produit IA de votre start-up

## Analyser les commentaires pour détecter les différents sujets d’insatisfaction

## Fichier csv détecter les sujets d’insatisfaction

### Importer la librairie

In [1]:
import tarfile
import numpy as np
import json
import pandas as pd
from sys import getsizeof

## Définir mes 2 jeux de données(textuelles)

Pour créer mes deux jeux de données, je vais tout d’abord décompresser le dossier 'yelp_dataset.tar’ en utilisant la méthode tarfile.open qui m’a permis de sauvegarder automatiquement mes 5 fichiers Json dans mon répertoire de travail, ensuite j’ai intégré dans ma dataframe nommé df_affaire mon premier fichier Json nommé 'yelp_academic_dataset_business.json’ en utilisant la mèthode read_json, à cette instant j’ai pu sélectionner uniquement la catégorie Restaurants dans mon premier jeu de données qui comporte 163 observations et 14 variables, sachant qu’il y a que 2 variables utiles qui sont ‘business_id’ et ‘categories’, je les ai gardé et j’ai supprimé les autres pour arrivé à 163 observations et 2 variables.  Ensuite pour récolter les données mon second fichier Json nommé 'yelp_academic_dataset_review.json’, du fait que le fichier est trop volumineux, j’ai du dans un premier lieu, définir une fonction nommé load_json2 qui intègre la dataframe en utilisant la mèthode read_json, plus la mèthode chunksize qui est une technique de reduction de dimension, plus un filtre qui me permet de récupérer uniquement les items qui m’interesse c’est à dire 'review_id', 'business_id', 'text’ et 'stars’. A cette instant, j’ai pus faire appelle à cette fonction nommé load_json2 sur mon fichier yelp_academic_dataset_review.json qui me permet de créer mon second jeu de données. 

### Extraction des différents fichiers JSon data_set

In [2]:
t = tarfile.open(name='yelp_dataset.tar', mode='r', fileobj=None, bufsize=10240)

In [3]:
t.extractall()
t.close()

### Définir l'emplacement des fichiers  

In [4]:
data_dir = "C:/Users/JK253/Documents/Ingenieur IA/P6/"

### Jeu de données entreprise fourni

In [5]:
df_affaire = pd.read_json('yelp_academic_dataset_business.json', lines=True)

In [6]:
df_affaire = df_affaire[df_affaire.categories == 'Restaurants']
df_affaire

business_id                             name  \
2725    pYDcpUeGAJHzUsE3l8vEtw                The Bar Next Door   
2887    37hDKzA07E3pmwCWczdi_w                           Mr Wok   
3991    9O5wLi4lUoa5GRP5BaAuVQ                             Figs   
9843    5SNSoEMZpMgg0D_vfaGL5w                 Boston Chowda Co   
10442   xWQl5VqVYIeZ5WYhGi57yA                         Finian's   
...                        ...                              ...   
154996  SKAABbhPfGtfS-PlmXc4SA  Sops On Ellis Loaf & Kettle Inc   
155296  hq-LM9NVixkrJDSGY2LDmA          Big Bully's BBQ Burgers   
158780  W3x6BBb090odXMA1iKCZLg        Big Easy Cajun Burlington   
158958  wmb4ZrHo5D5cxT9-sRjiVA                     RJ Crowley's   
159730  RnOCZCdbaI7oMjuRT2JqNQ                 Soho Bar & Grill   

                       address         city state postal_code   latitude  \
2725        5571 N Hamilton Rd     Columbus    OH       43230  40.070196   
2887    2050 Lawrenceville Hwy      Decatur    GA       30033  33.808683   
3991             92 Central St    Wellesley    MA       02481  42.295739   
9843                 36 JFK St    Cambridge    MA       02138  42.372650   
10442          1657 Hancock St       Quincy    MA       02169  42.245605   
...                        ...          ...   ...         ...        ...   
154996            170 Ellis St      Atlanta    GA       30303  33.758387   
155296    55 Dodge St, Ste 200      Beverly    MA       01915  42.575825   
158780       75 Middlesex Tpke   Burlington    MA       01803  42.482938   
158958      1391 Washington St  West Newton    MA       02465  42.348452   
159730       7090 Lougheed Hwy      Burnaby    BC     V5A 1W2  49.258204   

         longitude  stars  review_count  is_open  \
2725    -82.862670    3.5             5        0   
2887    -84.276401    3.0             8        0   
3991    -71.297635    4.0            11        0   
9843    -71.119733    2.5             6        0   
10442   -71.000567    2.5            15        0   
...            ...    ...           ...      ...   
154996  -84.382227    4.5             8        0   
155296  -70.889550    2.5             5        0   
158780  -71.213472    1.5             6        1   
158958  -71.229788    3.5             6        0   
159730 -122.956074    2.0             5        0   

                                               attributes   categories  \
2725    {'BusinessParking': '{'garage': False, 'street...  Restaurants   
2887    {'Ambience': '{'romantic': False, 'intimate': ...  Restaurants   
3991    {'BusinessParking': '{'garage': False, 'street...  Restaurants   
9843    {'RestaurantsTakeOut': 'True', 'RestaurantsPri...  Restaurants   
10442   {'GoodForKids': 'True', 'RestaurantsGoodForGro...  Restaurants   
...                                                   ...          ...   
154996  {'BusinessParking': '{'garage': False, 'street...  Restaurants   
155296  {'OutdoorSeating': 'False', 'GoodForKids': 'Tr...  Restaurants   
158780  {'BusinessAcceptsCreditCards': 'True', 'Restau...  Restaurants   
158958  {'BusinessParking': '{'garage': False, 'street...  Restaurants   
159730  {'Alcohol': 'u'full_bar'', 'RestaurantsDeliver...  Restaurants   

                                                    hours  
2725                                                 None  
2887                                                 None  
3991                                                 None  
9843                                                 None  
10442                                                None  
...                                                   ...  
154996  {'Monday': '11:0-14:30', 'Tuesday': '11:0-14:3...  
155296                                               None  
158780                                               None  
158958                                               None  
159730                                               None  

[163 rows x 14 columns]

In [7]:
df_affaire = df_affaire.drop(columns=['name', 'address', 'city', 'state','postal_code', 'latitude', 'longitude', 'stars',
                                     'review_count', 'is_open', 'attributes', 'hours'])

### Jeu de données avis des clients fourni

In [8]:
def load_json2(jpath):
    data_iterator = pd.read_json(jpath, chunksize=100000, lines=True)
    chunk_list = []
    for data_chunk in data_iterator:
        filtered_chunk = data_chunk.filter(items=['review_id', 'business_id', 'text', 'stars'])
        chunk_list.append(filtered_chunk)
        print('chunk appended')
        print()

    df = pd.concat(chunk_list)
    return df

In [9]:
df_1 = load_json2('yelp_academic_dataset_review.json')
print('Nombre de lignes et de colonnes: ', df_1.shape)
df_1.head()

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk appended

chunk ap

review_id             business_id  \
0  lWC-xP3rd6obsecCYsGZRg  buF9druCkbuXLX526sGELQ   
1  8bFej1QE5LXp4O05qjGqXA  RA4V8pr014UyUbDvI-LW2A   
2  NDhkzczKjLshODbqDoNLSg  _sS2LBIGNT5NQb6PD1Vtjw   
3  T5fAqjjFooT4V0OeZyuk1w  0AzLzHfOJgL7ROwhdww2ew   
4  sjm_uUcQVxab_EeLCqsYLg  8zehGz9jnxPqXtOc7KaJxA   

                                                text  stars  
0  Apparently Prides Osteria had a rough summer a...      4  
1  This store is pretty good. Not as great as Wal...      4  
2  I called WVM on the recommendation of a couple...      5  
3  I've stayed at many Marriott and Renaissance M...      2  
4  The food is always great here. The service fro...      4

## Créer un seul jeu de données d’avis d’insatisfaction client

Pour créer un seul jeu de données d’avis d’insatisfaction client, j’ai à ma disposition 2 jeux de données créé précédemment, plus le jeu de données importer d’API yelp, j’ai tout d’abord assembler mes 2 premiers jeux de données en utilisant la mèthode merge par l’intermédiaire de 'business_id’, et pour les 2 jeux restant par la mèthode concat, ensuite pour extraire les avis d’insatisfaction client, j’ai utiliser la variable ‘stars’ correspondant au note des clients en lui disant que toute les valeurs inférieurs à 3 correspond à des avis d’insatisfactions client sachant que la valeurs fourni par les clients varies de 0 à 5.

### Assenbler les 2 jeux de données mise à disposition

In [10]:
df_ext= pd.merge(df_affaire, df_1, how='inner', on=['business_id'])
df_ext

business_id   categories               review_id  \
0     pYDcpUeGAJHzUsE3l8vEtw  Restaurants  zRP7jDTIjze6eQU9L-CN1w   
1     pYDcpUeGAJHzUsE3l8vEtw  Restaurants  4RvDPUM8PmZwMiorL3kx7w   
2     pYDcpUeGAJHzUsE3l8vEtw  Restaurants  GS4ahbSI5VBa89QnhlaiCw   
3     pYDcpUeGAJHzUsE3l8vEtw  Restaurants  yqLSZmO1fKNecjJQz915Sw   
4     pYDcpUeGAJHzUsE3l8vEtw  Restaurants  rHsrbRNcIwTSpf5rFoP7RQ   
...                      ...          ...                     ...   
1975  RnOCZCdbaI7oMjuRT2JqNQ  Restaurants  bp1OwUjrMtzKdJUNM1tkBQ   
1976  RnOCZCdbaI7oMjuRT2JqNQ  Restaurants  run6dRX5KHw5dXZdCR_b8A   
1977  RnOCZCdbaI7oMjuRT2JqNQ  Restaurants  eYrNrgjDz8rbp2iJcY6Tow   
1978  RnOCZCdbaI7oMjuRT2JqNQ  Restaurants  PJtB6fpURI1EBsH2HO9nFg   
1979  RnOCZCdbaI7oMjuRT2JqNQ  Restaurants  DYLBFjOVOVpNJPjm1BGVKA   

                                                   text  stars  
0     The re-invention of the old Outside Corner, Th...      5  
1     The owner Steve  really cares about his patron...      4  
2     *yawn*  This is the forgettable sort of bar th...      2  
3     The Bar Next door is lit Door to where my girl...      3  
4     The Bar Next Door is a welcome reincarnation/n...      4  
...                                                 ...    ...  
1975  the prawn appetizer about 4 years ago was hell...      1  
1976  They ran out of wings and told us after we wai...      1  
1977  This place is just fine. We came to watch UFC ...      3  
1978  It appears to be closed down. Went at dinner t...      1  
1979  Good wings, good prices, and the music is at a...      3  

[1980 rows x 5 columns]

### Jeu de données avis des clients importer sur l'API Yelp

In [11]:
df_2 = pd.read_csv(data_dir + 'P6_01_fichiercsv_01.csv')
df_2.head()

text  stars
0  This place was awesome for lunch. We shared a ...      5
1  Absolutely the best Fish and Chips in London. ...      5
2  Enjoyed all our starters and the traditional f...      4
3  Amazing, I was shocked at how delicious this p...      5
4  I really hate to say OMG because I'm really to...      5

### Assembler le nouveau jeu de données

In [12]:
data=pd.concat([df_ext,df_2], axis=0)

In [13]:
data

business_id   categories               review_id  \
0     pYDcpUeGAJHzUsE3l8vEtw  Restaurants  zRP7jDTIjze6eQU9L-CN1w   
1     pYDcpUeGAJHzUsE3l8vEtw  Restaurants  4RvDPUM8PmZwMiorL3kx7w   
2     pYDcpUeGAJHzUsE3l8vEtw  Restaurants  GS4ahbSI5VBa89QnhlaiCw   
3     pYDcpUeGAJHzUsE3l8vEtw  Restaurants  yqLSZmO1fKNecjJQz915Sw   
4     pYDcpUeGAJHzUsE3l8vEtw  Restaurants  rHsrbRNcIwTSpf5rFoP7RQ   
...                      ...          ...                     ...   
2026                     NaN          NaN                     NaN   
2027                     NaN          NaN                     NaN   
2028                     NaN          NaN                     NaN   
2029                     NaN          NaN                     NaN   
2030                     NaN          NaN                     NaN   

                                                   text  stars  
0     The re-invention of the old Outside Corner, Th...      5  
1     The owner Steve  really cares about his patron...      4  
2     *yawn*  This is the forgettable sort of bar th...      2  
3     The Bar Next door is lit Door to where my girl...      3  
4     The Bar Next Door is a welcome reincarnation/n...      4  
...                                                 ...    ...  
2026  Went with my partner for dinner and was treate...      1  
2027  The food is fantastic and the restaurant is ju...      5  
2028  We came here for dim sum since I haven't had d...      4  
2029  We came here on a weekend, and had a pretty ba...      2  
2030  The service here was helpful and welcoming and...      5  

[4011 rows x 5 columns]

### Sélection uniquement des avis d'insatisfactions

In [14]:
data = data[data.stars < 3]
print('Nombre de lignes et de colonnes: ', data.shape)
data.head()

Nombre de lignes et de colonnes:  (613, 5)


business_id   categories               review_id  \
2   pYDcpUeGAJHzUsE3l8vEtw  Restaurants  GS4ahbSI5VBa89QnhlaiCw   
7   37hDKzA07E3pmwCWczdi_w  Restaurants  GTS9Lbh4ySSdM4ekBzjemw   
8   37hDKzA07E3pmwCWczdi_w  Restaurants  T_LHHdDIaw8DZr3PoICDlg   
22  9O5wLi4lUoa5GRP5BaAuVQ  Restaurants  WuRTnKy8t_6RGKsWy85uXw   
24  5SNSoEMZpMgg0D_vfaGL5w  Restaurants  6RdcVUQnOfUTk06igP_3ZQ   

                                                 text  stars  
2   *yawn*  This is the forgettable sort of bar th...      2  
7   Ordered all veggies, because of my clean eatin...      1  
8   I used to love this place. I ordered the honey...      2  
22  When I was at Babson, there was a big controve...      1  
24  We ended up at the location in Cambridge, MA j...      2

### Supprimer 3 variables

In [15]:
data = data.drop(columns=['business_id', 'review_id','categories'])
data

text  stars
2     *yawn*  This is the forgettable sort of bar th...      2
7     Ordered all veggies, because of my clean eatin...      1
8     I used to love this place. I ordered the honey...      2
22    When I was at Babson, there was a big controve...      1
24    We ended up at the location in Cambridge, MA j...      2
...                                                 ...    ...
1948  I ate here alone and felt I was an inconvenien...      2
1958  Extremely overpriced tourist trap. The portion...      1
1974  Not knowing a UK bacon sandwich is ONLY a thin...      2
2026  Went with my partner for dinner and was treate...      1
2029  We came here on a weekend, and had a pretty ba...      2

[613 rows x 2 columns]

## Création du fichier csv des Avis Insatisfactions clients

In [16]:
data.to_csv('P6_01_fichiercsv_02.csv', index = False)

Par la suite j’ai garder uniquement 2 variables qui sont ‘text et ‘stars’, mon nouveau jeu de données comporte 613 documents et 2 variables que j’ai enregistrer en fichier CSV nommé 'P6_01_fichiercsv_02.csv’.